In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [3]:
customers = pd.read_csv("../data/raw/olist dataset/customers.csv")

In [4]:
customers.head()

,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
0,06b8999e2fba1a1fbc88172c00ba8bc7,861eff4711a542e4b93843c6dd7febb0,14409,franca,SP
1,18955e83d337fd6b2def6b18a428ac77,290c77bc529b7ac935b93aa66c333dc3,9790,sao bernardo do campo,SP
2,4e7b3e00288586ebd08712fdd0374a03,060e732b5b29e8181a18229c7b0b2b5e,1151,sao paulo,SP
3,b2b6027bc5c5109e529d4dc6358b12c3,259dac757896d24d7702b9acbbff3f3c,8775,mogi das cruzes,SP
4,4f2d8ab171c80ec8364f7c12e35b23ad,345ecd01c38d18a9036ed96c73b8d066,13056,campinas,SP


In [5]:
customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   customer_id               99441 non-null  object
 1   customer_unique_id        99441 non-null  object
 2   customer_zip_code_prefix  99441 non-null  int64 
 3   customer_city             99441 non-null  object
 4   customer_state            99441 non-null  object
dtypes: int64(1), object(4)
memory usage: 3.8+ MB


In [6]:
customers.shape

(99441, 5)

In [7]:
customers.isnull().sum()

customer_id                 0
customer_unique_id          0
customer_zip_code_prefix    0
customer_city               0
customer_state              0
dtype: int64

In [8]:
customers.duplicated().sum()

0

# Customers Dataset Profiling

## Dataset Summary

- Rows: 99,441
- Columns: 5
- Missing Values: 0
- Duplicate Rows: 0

### Key Observations

- Total Records: 99,441
- Missing Values: None
- Duplicate Records: None
- Primary Key: customer_id
- Data Quality: Excellent
- Ready for PostgreSQL import.

## Primary Key

customer_id is the primary key.

Reason:
Each customer_id uniquely identifies one customer record.

## Business Understanding

This dataset stores customer location information.

Business questions:

- Which states generate the highest revenue?
- Which cities have the most customers?
- Which regions have low customer penetration?

# Orders Dataset Profiling

In [9]:
orders = pd.read_csv("../data/raw/olist dataset/orders.csv")

In [10]:
orders.head()

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18 00:00:00
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13 00:00:00
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04 00:00:00
3,949d5b44dbf5de918fe9c16f97b45f8a,f88197465ea7920adcdbec7375364d82,delivered,2017-11-18 19:28:06,2017-11-18 19:45:59,2017-11-22 13:39:59,2017-12-02 00:28:42,2017-12-15 00:00:00
4,ad21c59c0840e6cb83a9ceb5573f8159,8ab97904e6daea8866dbdbc4fb7aad2c,delivered,2018-02-13 21:18:39,2018-02-13 22:20:29,2018-02-14 19:46:34,2018-02-16 18:17:02,2018-02-26 00:00:00


In [11]:
orders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   order_id                       99441 non-null  object
 1   customer_id                    99441 non-null  object
 2   order_status                   99441 non-null  object
 3   order_purchase_timestamp       99441 non-null  object
 4   order_approved_at              99281 non-null  object
 5   order_delivered_carrier_date   97658 non-null  object
 6   order_delivered_customer_date  96476 non-null  object
 7   order_estimated_delivery_date  99441 non-null  object
dtypes: object(8)
memory usage: 6.1+ MB


In [12]:
orders.shape

(99441, 8)

In [13]:
orders.isnull().sum()

order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
dtype: int64

In [14]:
orders.duplicated().sum()

0

## Dataset Summary

- Rows: 99,441
- Columns: 8
- Duplicate Records: 0

Missing Values

- order_approved_at : 160
- order_delivered_carrier_date : 1,783
- order_delivered_customer_date : 2,965

Observation

Missing values are business-related rather than data quality issues. They indicate orders that may be pending, cancelled, or not yet delivered.

## Primary Key

order_id is the primary key.

Each order is uniquely identified using order_id.

## Business Understanding

This dataset contains the complete lifecycle of customer orders.

Business Questions

- How many orders were delivered?
- Which order statuses occur most frequently?
- What is the average delivery time?
- How many cancelled orders exist?
- Which months generate the highest order volume?

## Key Observations

- Data Quality: Good
- Duplicate Records: None
- Missing Values are business-related
- Suitable for PostgreSQL import

# Order Items Dataset Profiling

In [21]:
orderitems = pd.read_csv("../data/raw/olist dataset/orderitems.csv")

In [22]:
orders.head()

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18 00:00:00
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13 00:00:00
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04 00:00:00
3,949d5b44dbf5de918fe9c16f97b45f8a,f88197465ea7920adcdbec7375364d82,delivered,2017-11-18 19:28:06,2017-11-18 19:45:59,2017-11-22 13:39:59,2017-12-02 00:28:42,2017-12-15 00:00:00
4,ad21c59c0840e6cb83a9ceb5573f8159,8ab97904e6daea8866dbdbc4fb7aad2c,delivered,2018-02-13 21:18:39,2018-02-13 22:20:29,2018-02-14 19:46:34,2018-02-16 18:17:02,2018-02-26 00:00:00


In [23]:
orderitems.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   order_id             112650 non-null  object 
 1   order_item_id        112650 non-null  int64  
 2   product_id           112650 non-null  object 
 3   seller_id            112650 non-null  object 
 4   shipping_limit_date  112650 non-null  object 
 5   price                112650 non-null  float64
 6   freight_value        112650 non-null  float64
dtypes: float64(2), int64(1), object(4)
memory usage: 6.0+ MB


In [24]:
orderitems.shape

(112650, 7)

In [25]:
orderitems.isnull().sum()

order_id               0
order_item_id          0
product_id             0
seller_id              0
shipping_limit_date    0
price                  0
freight_value          0
dtype: int64

In [26]:
orderitems.duplicated().sum()

0

## Dataset Summary

- Rows: 112,650
- Columns: 7
- Missing Values: 0
- Duplicate Rows: 0

Observation:

The order items dataset is clean and does not require preprocessing.

## Primary Key

There is no single primary key.

Composite Primary Key:

- order_id
- order_item_id

Reason:

A single order can contain multiple products. Therefore, order_id alone is not unique. The combination of order_id and order_item_id uniquely identifies each order item.

## Foreign Keys

- order_id → Orders
- product_id → Products
- seller_id → Sellers

## Business Understanding

Each row represents one product purchased within an order.

Business Questions

- Which products generate the highest revenue?
- Which sellers contribute the highest sales?
- Which products have the highest freight cost?
- Which orders contain multiple products?
- What is the average item price?

## AI Business Use Cases

- Product Performance Analysis
- Seller Performance Analysis
- Revenue Insights
- Freight Cost Optimization
- Executive Sales Copilot

# Products Dataset Profiling

In [27]:
products = pd.read_csv("../data/raw/olist dataset/products.csv")

In [28]:
products.head()

,product_id,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
0,1e9e8ef04dbcff4541ed26657ea517e5,perfumaria,40.0,287.0,1.0,225.0,16.0,10.0,14.0
1,3aa071139cb16b67ca9e5dea641aaa2f,artes,44.0,276.0,1.0,1000.0,30.0,18.0,20.0
2,96bd76ec8810374ed1b65e291975717f,esporte_lazer,46.0,250.0,1.0,154.0,18.0,9.0,15.0
3,cef67bcfe19066a932b7673e239eb23d,bebes,27.0,261.0,1.0,371.0,26.0,4.0,26.0
4,9dc1a7de274444849c219cff195d0b71,utilidades_domesticas,37.0,402.0,4.0,625.0,20.0,17.0,13.0


In [32]:
products.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32951 entries, 0 to 32950
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   product_id                  32951 non-null  object 
 1   product_category_name       32341 non-null  object 
 2   product_name_lenght         32341 non-null  float64
 3   product_description_lenght  32341 non-null  float64
 4   product_photos_qty          32341 non-null  float64
 5   product_weight_g            32949 non-null  float64
 6   product_length_cm           32949 non-null  float64
 7   product_height_cm           32949 non-null  float64
 8   product_width_cm            32949 non-null  float64
dtypes: float64(7), object(2)
memory usage: 2.3+ MB


In [31]:
products.shape

(32951, 9)

In [30]:
products.isnull().sum()

product_id                      0
product_category_name         610
product_name_lenght           610
product_description_lenght    610
product_photos_qty            610
product_weight_g                2
product_length_cm               2
product_height_cm               2
product_width_cm                2
dtype: int64

In [29]:
products.duplicated().sum()

0

## Dataset Summary

- Rows: 32,951
- Columns: 9
- Missing Values:
  - product_category_name: 610
  - product_name_lenght: 610
  - product_description_lenght: 610
  - product_photos_qty: 610
  - product_weight_g: 2
  - product_length_cm: 2
  - product_height_cm: 2
  - product_width_cm: 2
- Duplicate Rows: 0

Observation:

The dataset contains missing values mainly in product metadata and category-related fields. These records will require preprocessing before analysis.

## Primary Key

product_id is the primary key.

Reason:

Each product is uniquely identified by product_id.

## Foreign Key

product_category_name → Categories

## Business Understanding

Each row represents one unique product listed on the Olist marketplace.

Business Questions

- Which product categories generate the highest revenue?
- Which product categories are sold most frequently?
- What is the average product weight and dimensions?
- Which products require higher shipping costs?
- Which product categories have incomplete information?

## AI Business Use Cases

- Product Performance Analysis
- Product Recommendation System
- Category Performance Dashboard
- Inventory Optimization
- Product Catalog Analytics

# Payments Dataset Profiling

In [33]:
payments = pd.read_csv("../data/raw/olist dataset/payments.csv")

In [34]:
payments.head()

,order_id,payment_sequential,payment_type,payment_installments,payment_value
0,b81ef226f3fe1789b1e8b2acac839d17,1,credit_card,8,99.33
1,a9810da82917af2d9aefd1278f1dcfa0,1,credit_card,1,24.39
2,25e8ea4e93396b6fa0d3dd708e76c1bd,1,credit_card,1,65.71
3,ba78997921bbcdc1373bb41e913ab953,1,credit_card,8,107.78
4,42fdf880ba16b47b59251dd489d4441a,1,credit_card,2,128.45


In [35]:
payments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103886 entries, 0 to 103885
Data columns (total 5 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   order_id              103886 non-null  object 
 1   payment_sequential    103886 non-null  int64  
 2   payment_type          103886 non-null  object 
 3   payment_installments  103886 non-null  int64  
 4   payment_value         103886 non-null  float64
dtypes: float64(1), int64(2), object(2)
memory usage: 4.0+ MB


In [36]:
payments.shape

(103886, 5)

In [37]:
payments.isnull().sum()

order_id                0
payment_sequential      0
payment_type            0
payment_installments    0
payment_value           0
dtype: int64

In [62]:
payments.duplicated().sum()

0

## Dataset Summary

- Rows: 103,886
- Columns: 5
- Missing Values: 0
- Duplicate Rows: 0

Observation:

The payments dataset is clean and complete. It contains transaction-level payment information for each order and does not require preprocessing.

## Primary Key

There is no single primary key.

Composite Primary Key:

- order_id
- payment_sequential

Reason:

An order may contain multiple payment records, making the combination of order_id and payment_sequential unique.

## Foreign Key

order_id → Orders

## Business Understanding

Each row represents a payment transaction made for an order.

Business Questions

- Which payment method is used most frequently?
- What is the average payment value?
- How many installment payments are used?
- Which payment methods generate the highest revenue?
- What is the distribution of payment values?

## AI Business Use Cases

- Payment Method Analysis
- Installment Behavior Analysis
- Revenue Analytics
- Customer Payment Preference Analysis
- Financial Dashboard

# Reviews Dataset Profiling

In [39]:
reviews = pd.read_csv("../data/raw/olist dataset/reviews.csv")

In [40]:
reviews.head()

,review_id,order_id,review_score,review_comment_title,review_comment_message,review_creation_date,review_answer_timestamp
0,7bc2406110b926393aa56f80a40eba40,73fc7af87114b39712e6da79b0a377eb,4,NaN,NaN,2018-01-18 00:00:00,2018-01-18 21:46:59
1,80e641a11e56f04c1ad469d5645fdfde,a548910a1c6147796b98fdf73dbeba33,5,NaN,NaN,2018-03-10 00:00:00,2018-03-11 03:05:13
2,228ce5500dc1d8e020d8d1322874b6f0,f9e4b658b201a9f2ecdecbb34bed034b,5,NaN,NaN,2018-02-17 00:00:00,2018-02-18 14:36:24
3,e64fb393e7b32834bb789ff8bb30750e,658677c97b385a9be170737859d3511b,5,NaN,Recebi bem antes do prazo estipulado.,2017-04-21 00:00:00,2017-04-21 22:02:06
4,f7c4243c7fe1938f181bec41a392bdeb,8e6bfb81e283fa7e4f11123a3fb894f1,5,NaN,Parabéns lojas lannister adorei comprar pela I...,2018-03-01 00:00:00,2018-03-02 10:26:53


In [41]:
reviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99224 entries, 0 to 99223
Data columns (total 7 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   review_id                99224 non-null  object
 1   order_id                 99224 non-null  object
 2   review_score             99224 non-null  int64 
 3   review_comment_title     11568 non-null  object
 4   review_comment_message   40977 non-null  object
 5   review_creation_date     99224 non-null  object
 6   review_answer_timestamp  99224 non-null  object
dtypes: int64(1), object(6)
memory usage: 5.3+ MB


In [63]:
reviews.shape

(99224, 7)

In [42]:
reviews.isnull().sum()

review_id                      0
order_id                       0
review_score                   0
review_comment_title       87656
review_comment_message     58247
review_creation_date           0
review_answer_timestamp        0
dtype: int64

In [43]:
reviews.duplicated().sum()

0

## Dataset Summary

- Rows: 99,224
- Columns: 7
- Missing Values:
  - review_comment_title: 87,656
  - review_comment_message: 58,247
- Duplicate Rows: 0

### Observation

The reviews dataset contains customer ratings and optional written feedback. Missing values in review titles and messages are expected because many customers submitted only a rating without providing textual comments. Therefore, these missing values are not considered data quality issues.

## Primary Key

**Primary Key:** `review_id`

### Reason

Each review is uniquely identified by its `review_id`, ensuring that every customer review record is distinct.

## Foreign Key

- `order_id` → Orders Dataset (`order_id`)

## Business Understanding

Each row represents a customer's review submitted after receiving an order.

### Business Questions

- What is the average customer review score?
- Which products receive the highest and lowest ratings?
- Which sellers receive poor customer feedback?
- How many customers leave written reviews?
- What factors influence customer satisfaction?

## AI Business Use Cases

- Customer Sentiment Analysis
- Customer Satisfaction Dashboard
- Product Rating Analysis
- Seller Performance Monitoring
- Review Text Analytics
- Customer Experience Insights

# Sellers Dataset Profiling

In [44]:
sellers = pd.read_csv("../data/raw/olist dataset/sellers.csv")

In [45]:
sellers.head()

,seller_id,seller_zip_code_prefix,seller_city,seller_state
0,3442f8959a84dea7ee197c632cb2df15,13023,campinas,SP
1,d1b65fc7debc3361ea86b5f14c68d2e2,13844,mogi guacu,SP
2,ce3ad9de960102d0677a81f5d0bb7b2d,20031,rio de janeiro,RJ
3,c0f3eea2e14555b6faeea3dd58c1b1c3,4195,sao paulo,SP
4,51a04a8a6bdcb23deccc82b0b80742cf,12914,braganca paulista,SP


In [46]:
sellers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3095 entries, 0 to 3094
Data columns (total 4 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   seller_id               3095 non-null   object
 1   seller_zip_code_prefix  3095 non-null   int64 
 2   seller_city             3095 non-null   object
 3   seller_state            3095 non-null   object
dtypes: int64(1), object(3)
memory usage: 96.8+ KB


In [47]:
sellers.shape

(3095, 4)

In [48]:
sellers.isnull().sum()

seller_id                 0
seller_zip_code_prefix    0
seller_city               0
seller_state              0
dtype: int64

In [49]:
sellers.duplicated().sum()

0

## Dataset Summary

- Rows: 3,095
- Columns: 4
- Missing Values: 0
- Duplicate Rows: 0

Observation:

The sellers dataset is clean and complete without missing or duplicate records.

## Primary Key

seller_id is the primary key.

Reason:

Each seller is uniquely identified by seller_id.

## Foreign Key

seller_id → Order Items

## Business Understanding

Each row represents a seller registered on the marketplace.

Business Questions

- Which cities have the highest number of sellers?
- Which states contribute the most sellers?
- How are sellers geographically distributed?
- Which sellers generate the highest revenue?
- Which regions have low seller density?

## AI Business Use Cases

- Seller Performance Analysis
- Geographic Seller Analysis
- Marketplace Expansion Planning
- Seller Distribution Dashboard
- Regional Business Intelligence

# Categories Dataset Profiling

In [50]:
categories = pd.read_csv("../data/raw/olist dataset/categories.csv")

In [51]:
categories.head()

,product_category_name,product_category_name_english
0,beleza_saude,health_beauty
1,informatica_acessorios,computers_accessories
2,automotivo,auto
3,cama_mesa_banho,bed_bath_table
4,moveis_decoracao,furniture_decor


In [52]:
categories.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 71 entries, 0 to 70
Data columns (total 2 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   product_category_name          71 non-null     object
 1   product_category_name_english  71 non-null     object
dtypes: object(2)
memory usage: 1.2+ KB


In [53]:
categories.shape

(71, 2)

In [54]:
categories.isnull().sum()

product_category_name            0
product_category_name_english    0
dtype: int64

In [55]:
categories.duplicated().sum()

0

## Dataset Summary

- Rows: 71
- Columns: 2
- Missing Values: 0
- Duplicate Rows: 0

Observation:

The categories dataset serves as a lookup table for translating Portuguese product categories into English.

## Primary Key

product_category_name is the primary key.

Reason:

Each category appears only once.

## Foreign Key

product_category_name → Products

## Business Understanding

Each row maps a Portuguese product category to its English equivalent.

Business Questions

- Which product categories generate the highest sales?
- Which categories have the most products?
- Which categories contribute the highest revenue?
- Which categories receive the highest customer ratings?

## AI Business Use Cases

- Product Category Analysis
- Category Performance Dashboard
- Revenue Analysis by Category
- Product Recommendation
- Business Intelligence Reporting

# Geolocation Dataset Profiling

In [56]:
geolocation = pd.read_csv("../data/raw/olist dataset/geolocation.csv")

In [57]:
geolocation.head()

,geolocation_zip_code_prefix,geolocation_lat,geolocation_lng,geolocation_city,geolocation_state
0,1037,-23.545621,-46.639292,sao paulo,SP
1,1046,-23.546081,-46.644820,sao paulo,SP
2,1046,-23.546129,-46.642951,sao paulo,SP
3,1041,-23.544392,-46.639499,sao paulo,SP
4,1035,-23.541578,-46.641607,sao paulo,SP


In [58]:
geolocation.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000163 entries, 0 to 1000162
Data columns (total 5 columns):
 #   Column                       Non-Null Count    Dtype  
---  ------                       --------------    -----  
 0   geolocation_zip_code_prefix  1000163 non-null  int64  
 1   geolocation_lat              1000163 non-null  float64
 2   geolocation_lng              1000163 non-null  float64
 3   geolocation_city             1000163 non-null  object 
 4   geolocation_state            1000163 non-null  object 
dtypes: float64(2), int64(1), object(2)
memory usage: 38.2+ MB


In [59]:
geolocation.shape

(1000163, 5)

In [60]:
geolocation.isnull().sum()

geolocation_zip_code_prefix    0
geolocation_lat                0
geolocation_lng                0
geolocation_city               0
geolocation_state              0
dtype: int64

In [61]:
geolocation.duplicated().sum()

261831

## Dataset Summary

- Rows: 1,000,163
- Columns: 5
- Missing Values: 0
- Duplicate Rows: 261,831

Observation:

The dataset contains a large number of duplicate geolocation records because multiple customers and sellers can share the same ZIP code and geographic coordinates. These duplicates are expected and should not be removed without considering the business context.

## Primary Key

There is no primary key.

Reason:

Multiple records can share the same ZIP code and geographic coordinates.

## Foreign Key

geolocation_zip_code_prefix → Customers

geolocation_zip_code_prefix → Sellers

## Business Understanding

Each row represents a geographic location associated with a ZIP code.

Business Questions

- Where are most customers located?
- Where are sellers concentrated?
- Which regions generate the highest sales?
- What is the geographic coverage of the marketplace?
- Which locations have low seller availability?

## AI Business Use Cases

- Geographic Sales Analysis
- Customer Location Analysis
- Seller Distribution Mapping
- Logistics Optimization
- Regional Business Intelligence